File: Trie/208. Implement Trie (Prefix Tree)/208. Implement Trie (Prefix Tree).java
class Trie {
    class TrieNode {
        boolean isWord;
        TrieNode[] next;
        public TrieNode() {
            isWord = false;
            next = new TrieNode[26];
        }
    }
    TrieNode root;
    /** Initialize your data structure here. */
    public Trie() {
        root = new TrieNode();
    }
    
    /** Inserts a word into the trie. */
    public void insert(String word) {
        TrieNode node = root;
        for (int i = 0; i < word.length(); i++) {
            char c = word.charAt(i);
            if (node.next[c-'a'] == null) node.next[c-'a'] = new TrieNode();
            node = node.next[c-'a'];
        }
        node.isWord = true;
    }
    
    /** Returns if the word is in the trie. */
    public boolean search(String word) {
        TrieNode node = root;
        for (int i = 0; i < word.length(); i++) {
            char c = word.charAt(i);
            if (node.next[c-'a'] == null) return false;
            node = node.next[c-'a'];
        }
        return node.isWord;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    public boolean startsWith(String prefix) {
        TrieNode node = root;
        for (int i = 0; i < prefix.length(); i++) {
            char c = prefix.charAt(i);
            if (node.next[c-'a'] == null) return false;
            node = node.next[c-'a'];
        }
        return true;
    }
}

File: Trie/208. Implement Trie (Prefix Tree)/208. Implement Trie (Prefix Tree).py
class TrieNode:

    def __init__(self, letter=None):
        """
        # initialize TrieNode
        # it should have link to the downstream neighbirs
        # also a signal that checks whether current node is end of
        # a word
        """
        self.next={}
        self.isEnd=False
        
class Trie:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root=TrieNode()

    def insert(self, word: str) -> None:
        """
        Inserts a word into the trie.
        """
        head=self.root
        for c in word:
            if c not in head.next:
                head.next[c]=TrieNode()
            head=head.next[c]
        head.isEnd=True
        

    def search(self, word: str) -> bool:
        """
        Returns if the word is in the trie.
        """
        head=self.root
        for c in word:
            if c not in head.next:
                return False
            head=head.next[c]
        return head.isEnd
        

    def startsWith(self, prefix: str) -> bool:
        """
        Returns if there is any word in the trie that starts with the given prefix.
        """
        head=self.root
        for c in prefix:
            if c not in head.next:
                return False
            head=head.next[c]
        return True
